ddtrace/internal/buffer.py

```diff
@@ -0,0 +1,57 @@
+from collections import deque
+import threading
+
+from ddtrace.vendor import attr
+
+
+class BufferFull(Exception):
+    pass
+
+
+class BufferItemTooLarge(Exception):
+    pass
+
+
+@attr.s
+class TraceBuffer(object):
+
+    max_size = attr.ib(type=int)
+    max_item_size = attr.ib(type=int)
+    _size = attr.ib(init=False, type=int, default=0, repr=True)
+    _lock = attr.ib(init=False, factory=threading.Lock, repr=False)
+    _buffer = attr.ib(init=False, factory=deque, repr=False)
+
+    def __len__(self):
+        return len(self._buffer)
+
+    @property
+    def size(self):
+        with self._lock:
+            return self._size
+
+    def _clear(self):
+        self._buffer.clear()
+        self._size = 0
+
+    def put(self, item):
+        item_len = len(item)
+        if item_len > self.max_item_size or item_len > self.max_size:
+            raise BufferItemTooLarge()
```

Comment:
this is weird because if max_size is 10 then you can have 10 items of length 10 (so 100 items), but you cannot have 1 item of length 11 only (so 11 items)…

Reply:
This is intentional since it's the payload limit. Breaking trace payloads up results in broken metrics being calculated in the agent 😕